Bug/fix long pushes end up in invalid state

myieye opened the pull request:
If a request times out while `hg incoming` is running, the server assumes that `hg unbundle` has started, but it hasn't.
So, this PR introduces a new state (validate) to better keep track of what the next server-side step is and be more forgiving in that way.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/hgresume/12)
<!-- Reviewable:end -->

myieye committed:
Introduce verify state and prevent invalid transaction state.

myieye committed:
Fix unit tests

jasonleenaylor commented:
*[`api/src/BundleHelper.php` line 9 at r1](https://reviewable.io/reviews/sillsdev/hgresume/12#-NyQx1Ce28tryNNdpt3L:-NyQx1Ce28tryNNdpt3M:b-i3l51k) ([raw file](https://github.com/sillsdev/hgresume/blob/2e5a827030e7cffae91570b20cad5b6befdb551c/api/src/BundleHelper.php#L9)):*
> ```PHP
> 	const State_Downloading = 'Downloading';
> 	const State_Uploading   = 'Uploading';
> 	const State_Verify      = 'Verify';
> ```

Verify is an actual hg command, and that isn't what you are doing here so I'd recommend a different term. Pending maybe? Anyhow, something that isn't overloading a term in mercurial space. :smile: 



_Code quote:_
```PHP
'Verify'
```


<!-- Sent from Reviewable.io -->

megahirt commented on api/src/AsyncRunner.php:
```diff
@@ -64,17 +64,17 @@ public function cleanUp() {
 	/**
-	 * 
-	 * @throws AsyncRunnerException
+	 * Waits for up to 5s for the runner to complete
```
```suggestion
	 * Waits for up to 50s for the runner to complete
```
50 seconds?

myieye replied:
No, it's really only 5s.

I went for a low value, because:
- Chorus times out after 30 seconds.
- This is only part of the request processing.
- The last chunk will generally land here twice (i.e. 10s)
- It doesn't matter if the number is too low, it just means the client will poll a bit more often

megahirt replied:
ok sounds good.

megahirt commented on api/src/RestServer.php:
```diff
@@ -147,6 +147,12 @@ static function mapHgResponse($hgrCode) {
 				$codeString = 'INPROGRESS';
 				break;
+			case HgResumeResponse::TIMEOUT:
+				// 408 is not currently handled by the client, so it will be
```
we should specify the client version when we say it's not supported, and then update the comment when we know what client version does support it.

myieye replied:
This is for the current version. If we want to make the client keep polling we basically have to send it something it's not expecting, so I just tried to pick a status code that sort of made sense. It's suboptimal, because the client will retry the chunk and we'll just ignore it, but as far as I can tell there isn't a better solution for the current Chorus clients.

We could check if the client is accessing v4 and then return 202 + "INPROGRESS" instead. The client would have to be adapted for that of course.

But, this is for v3 right now.

megahirt replied:
This is a good enough solution and should hopefully fix v3.

Future work is to publish v4 (or maybe it's v5?) and fix some things in the chorus client as well that use only v5.

megahirt commented on api/test/HgResumeApi_Test.php:
```diff
@@ -279,7 +279,7 @@ function testPullBundleChunk_PullUntilFinished_AssembledBundleIsValid() {
 		$assembledBundle = '';
 		$bundleSize = 1; // initialize the bundleSize; it will be overwritten after the first API call
-		while (mb_strlen($assembledBundle) < $bundleSize) {
+		while (mb_strlen($assembledBundle, "8bit") < $bundleSize) {
```
I assume this is due to running this in a newer version of PHP, PHP 7.4?

myieye replied:
Honestely, I'm not quite sure. But almost all of the calls to `mb_strlen` **already** passed this as a paramter. And **all** the ones that **didn't** failed. So, it feels like this was an oversight.

megahirt replied:
ok sounds good, thanks for the explanation.

megahirt reviewed (changes requested):
great work!  This will end up being a very valuable fix I think.  I agree with @jasonleenaylor on naming the method verify.  What other name could we use?

Another challenge is adding a unit test that exercises this new feature.

myieye committed:
fixup! Introduce verify state and prevent invalid transaction state.

myieye committed:
Add failing unit test for hg incoming timeout

myieye committed:
Make fake-polling use tiny chunk-size

myieye commented:
@jasonleenaylor @megahirt This is ready for a real review now.

FYI: Confusingly enough I rearranged my commits, but I didn't change any existing ones.

- I changed "Verify" to "Validate". It should be a term that describes what step the server is at. A generic word like "Pending" doesn't do that.
- I made the comment about my status code not being handled by the client more descriptive so it's clear that in v3 we're intentionally sending an unknown status code back
- I've made it so that clients send very small chunks during the "fake polling" I've introduced
- I added a unit test that broke before my changes and now works (the 3 failing tests were already failing)
![image](https://github.com/sillsdev/hgresume/assets/12587509/78b30d72-0cda-48fb-b92a-49ec9bcaf735)

Here's what a 1GB Send/Receive looks like:
![image](https://github.com/sillsdev/hgresume/assets/12587509/aa72c02a-9abc-4aa8-a0fe-2f1353800941)

I somewhat unintentionally made the change that the client now always waits until the server is done unbundling. I think it would only retry once before and so might never see the result. I think this is better.

megahirt commented on api/src/HgRunner.php:
```diff
@@ -58,30 +54,60 @@ function unbundle($filepath) {
+	}
+
+	/**
+	 * @param AsyncRunner $asyncRunner
```
```suggestion
	 * @param string $filePath
```
fix up documentation

myieye replied:
Fixed

megahirt commented on api/src/AsyncRunner.php:
```diff
@@ -64,17 +64,17 @@ public function cleanUp() {
+	 * Waits for up to 5s for the runner to complete
+	 * @return boolean	True if complete, otherwise not complete yet
 	 */
 	public function synchronize() {
```
how about we rename synchronize() to be isComplete() or isFinished() ?  The previous implementation would ensure synchronization or throw an exception (or at least that was the intention).  The new functionality just waits up to 5 seconds for the job to complete before returning false.
```suggestion
	public function isComplete() {
```
We would need to change the references if we rename.

myieye replied:
Renamed it to `waitForIsComplete`. `isComplete` already exists. It's what this uses while it waits.

megahirt reviewed (approved):
I am happy with this PR, thank you!  I would like @jasonleenaylor to have a chance to review before this is merged, unless he is too busy.

I think changing the name from synchronize() to isComplete() reads better, since the function does something different now.

myieye committed:
Rename synchronize and prevent late unexpected failures in tests

myieye committed:
Fix indentation and unused variable.

myieye reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hgresume/12)* status: 0 of 9 files reviewed, 3 unresolved discussions (waiting on @jasonleenaylor and @megahirt)

___
*[`api/src/RestServer.php` line 151 at r1](https://reviewable.io/reviews/sillsdev/hgresume/12#gh-1609460843:-NyZoyzpBQpa-1Fodbpg:b-896fix) ([raw file](https://github.com/sillsdev/hgresume/blob/2e5a827030e7cffae91570b20cad5b6befdb551c/api/src/RestServer.php#L151)):*
<details><summary><i>Previously, megahirt (Christopher Hirt) wrote…</i></summary><blockquote>

This is a good enough solution and should hopefully fix v3.

Future work is to publish v4 (or maybe it's v5?) and fix some things in the chorus client as well that use only v5.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

hahn-kev commented:
This is really good work Tim, I'm excited to see it out in the wild.

hahn-kev commented:
talked with Jason and he's fine with us merging this.
Build is here: https://github.com/sillsdev/hgresume/actions/runs/9276571495
Released image is here: https://github.com/sillsdev/hgresume/pkgs/container/hgresume/222170135?tag=v2024-05-28
